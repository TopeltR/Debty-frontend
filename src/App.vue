<template>
    <router-view/>
</template>

<script>
    export default {
        name: "App",
    };
</script>

<style>
    #app {
        font-family: 'Avenir', Helvetica, Arial, sans-serif;
        -webkit-font-smoothing: antialiased;
        -moz-osx-font-smoothing: grayscale;
        color: #2c3e50;
    }

    .fa-shadow-hover:hover {
        filter: drop-shadow(0 0 1px #ba9fc9);
    }
</style>
