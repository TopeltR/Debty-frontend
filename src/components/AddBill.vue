<template>
    <b-modal ref="modal" class="col-12">
        <div class="col-12" slot="modal-title">
            {{pageTitle}}
            <font-awesome-icon v-if="!selectedBill" id="add-bill-info" icon='info-circle'
                               class="ml-1"></font-awesome-icon>
            <b-tooltip v-if="!selectedBill" target="add-bill-info"
                       title="Here you can add bill to the event, remember to add participants!"
                       placement="bottom"></b-tooltip>
        </div>
        <div class="col-10 offset-1">
            <b-row>
                <b-col cols="12" class='TM30px'>
                    <b-row>
                        <form @submit.prevent='' class='wide'>
                            <div class='form-group'>
                                <label for='title'>Title:</label>
                                <input type='text' class='form-control' v-model='bill.title' id='title' required=''
                                       placeholder='Enter bill title'>
                            </div>
                            <div class='form-group'>
                                <label for='description'>Description:</label>
                                <textarea id='description' class='form-control' v-model='bill.description' type='text'
                                          placeholder='Enter bill description' required=''></textarea>
                            </div>
                            <div class="form-group">
                                <label for="buyer">Buyer:</label>
                                <autocomplete id="buyer" v-model="buyer" :field="field" v-on:input="updateBuyer"
                                              :items="allPeople"
                                              :keyExtractor="getFullName"/>
                            </div>
                            <div class="form-group inline-form">
                                <label class="PR5px" for="sum">Sum:</label>
                                <input type="text" class="inline form-control skinny" id="sum" v-model="bill.sum"
                                       required="" v-on:change="displayNotMatchMessage">
                                <span class="ml-1">€</span>
                            </div>
                            <div class='form-group'>
                                <p v-if='addPersonState.people.length > 0'>Participants:</p>
                                <ul>
                                    <li v-for='person in addPersonState.people' v-bind:key='person.id'>
                                        <div class="form-group inline-form">
                                            <label class="PR5px col-9 col-md-6" :for="person.firstName">{{
                                                getFullName(person)
                                                }}</label>
                                            <input type="number" class="inline form-control skinny participation col-3"
                                                   :id="person.firstName" v-model="person.participation" required=""
                                                   v-on:change="displayNotMatchMessage">
                                            <span class="ml-1 col-1">€</span>
                                        </div>
                                    </li>
                                </ul>
                                <label for='user'>Add participants:</label>
                                <add-person id="user" :state="addPersonState"/>
                                <p class="red MT29px" v-bind:style='{display: notMatchDisplayProperty}'>Sum and
                                    participations
                                    don't match!</p>
                                <p class="red MT29px" v-bind:style="{display: errorDisplayProperty}">Not all required
                                    fields filled</p>
                            </div>
                        </form>
                    </b-row>
                </b-col>
            </b-row>
        </div>
        <div class="col-12" slot="modal-footer">
            <b-row>
                <b-col cols="6">
                    <b-button class="wide" variant="secondary" v-on:click="cancel">Cancel</b-button>
                </b-col>
                <b-col cols="6">
                    <b-button class="wide" variant="primary" v-on:click="save">Save</b-button>
                </b-col>
            </b-row>
        </div>
    </b-modal>
</template>

<script>
    import AddPerson from '@/components/AddPerson.vue';
    import Autocomplete from "./Autocomplete";
    import userStore from "../stores/UserStore";

    export default {
        name: 'AddBill',
        components: {Autocomplete, AddPerson},
        props: {
            state: {
                type: Object,
                showing: {
                    type: Boolean,
                },
                default: () => ({
                    showing: false,
                }),
            },
            selectedBill: {
                title: {type: String, default: ''},
                description: {type: String, default: ''},
                sum: {type: Number, default: 0},
                buyer: {type: Object, default: () => ({})},
                people: {type: Array, default: () => ([])},
                billPayments: {type: Array, default: () => ([])},
                creator: {type: Object},
            },
            event: {
                type: Object,
            },
        },
        watch: {
            state: {
                deep: true,
                handler(state) {
                    if (state.showing) {
                        this.$refs.modal.show();
                    } else {
                        this.$refs.modal.hide();
                    }
                },
            },
            selectedBill() {
                this.setSelectedBillInfo();
            },
        },
        async mounted() {
            this.user = await userStore.getUser();
            const data = await this.$http.get('/contact/id/' + this.user.id);
            data.data.push(this.user);
            this.allPeople = data.data.map((u) => {
                u.participation = 0;
                return u;
            });
            // Copy, can't be same
            this.addPersonState.allPeople = this.allPeople.filter((u) => !this.bill.people.includes(u));
            this.addPersonState.people = this.bill.people;

            this.setSelectedBillInfo();
        },
        data: () => ({
            initialBill: {},
            pageTitle: 'Add bill',
            user: {},
            field: {value: ''},
            allPeople: [],
            buyer: {},
            participationChanged: false,
            billChanged: false,
            notMatchDisplayProperty: 'none',
            errorDisplayProperty: 'none',
            bill: {
                title: '',
                description: '',
                sum: 0,
                buyer: {},
                creator: {},
                people: [],
                billPayments: [],
            },
            addPersonState: {
                allPeople: [],
                people: [],
            },
        }),
        methods: {
            setSelectedBillInfo() {
                if (this.selectedBill) {
                    const bill = this.selectedBill;
                    for (const payment of bill.billPayments) {
                        const person = bill.people.find((p) => p.email === payment.person.email);
                        person.participation = payment.sum;
                    }

                    this.bill = {...bill};
                    if (!this.bill.people) {
                        this.bill.people = [];
                    }
                    this.initialBill = {...bill};
                    this.buyer = this.bill.buyer;
                    this.field.value = this.buyer.firstName + ' ' + this.buyer.lastName;
                    this.addPersonState.people = this.bill.people;
                }
            },
            updateBuyer() {
                if (this.buyer && this.buyer.firstName) {
                    this.addPersonState.people.removeElement(this.bill.buyer, (person) => person.email);
                    this.bill.buyer = this.buyer;
                    this.addPersonState.people.push(this.buyer);
                    this.buyer = {};
                }
            },
            getFullName(person) {
                return person.firstName + ' ' + person.lastName;
            },
            roundToTwoDecimalPoints(nr) {
                return Math.round(nr * 100) / 100;
            },
            displayNotMatchMessage() {
                const match =
                    this.roundToTwoDecimalPoints(this.addPersonState.people.map((u) => Number(u.participation))
                        .reduce((a, b) => this.roundToTwoDecimalPoints(a + b), 0)) === this.roundToTwoDecimalPoints(Number(this.bill.sum));
                if (!match) {
                    this.notMatchDisplayProperty = 'block';
                } else {
                    this.notMatchDisplayProperty = 'none';
                }
            },
            async save() {
                if (this.bill.title && this.bill.people.length > 0) {
                    this.errorDisplayProperty = 'none';
                    this.bill.people = this.addPersonState.people;

                    this.bill.billPayments = [];
                    for (const person of this.bill.people) {
                        this.bill.billPayments.push({
                            person,
                            sum: person.participation,
                        });
                    }

                    this.bill.creator = await userStore.getUser();

                    const response = await this.$http.post('/events/' + this.event.id + '/bills', this.bill);
                    this.event.bills = response.data.bills;

                    this.state.showing = true;
                    this.state.showing = false;
                } else {
                    this.errorDisplayProperty = 'block';
                }
            },
            cancel() {
                this.state.showing = true;
                this.state.showing = false;
            },
        },
    };
</script>

<style scoped>
    .wide {
        width: 100%;
    }

    .PR5px {
        padding-right: 5px;
    }

    .inline {
        display: inline;
    }

    .inline-form {
        white-space: nowrap;
    }

    .skinny {
        width: 70px;
    }

    .red {
        color: red;
    }

    .MT29px {
        margin-top: 29px;
    }
</style>
