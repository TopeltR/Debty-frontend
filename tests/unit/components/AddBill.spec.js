import AddBill from '@/components/AddBill';
import {mount} from "@vue/test-utils";
import userStore from '@/stores/UserStore';
import {Vue} from "vue-property-decorator";
import BootstrapVue from 'bootstrap-vue';
import App from "@/App";

describe('AddBill', () => {
    Vue.use(BootstrapVue);
    Vue.use(App);

    console.error = jest.fn();
    console.warn = jest.fn();

    userStore.getUser = jest.fn(() => Promise.resolve({id: 1}));

    describe('updateBuyer', () => {
        it('should add buyer into addPersonState.people and bill.buyer and empty buyer', () => {
            const addBill = mount(AddBill, {
                propsData: {
                    state: {},
                    selectedBill: {},
                    event: {}
                }
            });
            addBill.vm.$http = {get: jest.fn(() => Promise.resolve({data: []}))};
            addBill.vm.bill = {buyer: {}};
            const buyer = {firstName: 'first name'};
            addBill.vm.buyer = buyer;

            addBill.vm.updateBuyer();

            expect(addBill.vm.addPersonState.people.includes(buyer));
            expect(addBill.vm.buyer).toEqual({});
            expect(addBill.vm.bill.buyer).toEqual(buyer);
        })
    });

    describe('save', () => {
        it('should transform bill and post bill to /events/{event.id}/bills', async () => {
            const addBill = mount(AddBill, {
                propsData: {
                    state: {},
                    selectedBill: {},
                    event: {}
                }
            });
            const bill = {title: 'title', description: 'description', people: [{id: 1}], billPayments: []};
            addBill.vm.bill = bill;
            const person = {id: 1, participation: 10};
            addBill.vm.addPersonState.people = [person];
            addBill.vm.event.id = 5;
            const bills = [{id: 3}];
            addBill.vm.$http = {post: jest.fn(() => Promise.resolve({data: {bills}}))};

            await addBill.vm.save();

            expect(addBill.vm.$http.post).toHaveBeenCalledWith('/events/5/bills', {
                ...bill,
                people: [person],
                billPayments: [{person, sum: 10}]
            });
            expect(addBill.vm.event.bills).toEqual(bills);
            expect(addBill.vm.state.showing).toBe(false);
        });
    });

    it('should render', () => {
        const addBill = mount(AddBill, {
            propsData: {
                state: {},
                selectedBill: {},
                event: {}
            }
        });

        expect(addBill.html()).toMatchSnapshot();
    })
});
