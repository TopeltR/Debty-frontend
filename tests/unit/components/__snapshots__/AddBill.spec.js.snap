// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBill should render 1`] = `
<div class="col-12">
  <div style="position: absolute; z-index: 2000;">
    <div role="dialog" tabindex="-1" aria-hidden="true" class="modal fade" style="display: none;">
      <div class="modal-dialog modal-md">
        <div role="document" class="modal-content">
          <header class="modal-header">
            <h5 class="modal-title">
              <div class="col-12">
                Add bill
                <!---->
              </div>
            </h5><button type="button" aria-label="Close" class="close">×</button>
          </header>
          <div class="modal-body">
            <div class="col-10 offset-1">
              <div class="row">
                <div class="col-12">
                  <div class="row">
                    <form class="w-100">
                      <div class="form-group"><label for="title" class="font-weight-500">Title<span class="text-danger">&nbsp;*</span></label> <input type="text" id="title" required="required" placeholder="Enter bill title" minlength="1" maxlength="255" class="form-control"></div>
                      <div class="form-group"><label for="description" class="font-weight-500">Description
                          <!----></label> <textarea id="description" placeholder="Enter bill description" maxlength="255" class="form-control"></textarea></div>
                      <div class="form-group"><label for="buyer" class="font-weight-500">Buyer<span class="text-danger">&nbsp;*</span></label>
                        <div class="autocomplete" id="buyer" value="[object Object]"><input type="text" placeholder="" required="required" maxlength="255" class="form-control">
                          <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
                        </div>
                      </div>
                      <div class="form-group inline-form"><label for="sum" class="font-weight-500 pr-3">Sum<span class="text-danger">&nbsp;*</span></label> <input type="number" step="0.01" id="sum" required="required" min="0" maxlength="255" class="d-inline form-control skinny"> <span class="ml-1">€</span>
                        <!---->
                      </div>
                      <div class="form-group">
                        <!---->
                        <!----> <label for="user" class="font-weight-500">Add participants
                          <!----></label>
                        <div class="row" id="user">
                          <div col-md="9" class="col-8">
                            <div class="autocomplete" id="user" value="[object Object]"><input type="text" placeholder="Name" maxlength="255" class="form-control">
                              <ul id="autocomplete-results" class="autocomplete-results" style="display: none;"></ul>
                            </div>
                          </div>
                          <div col-md="3" class="col-4"><button type="button" class="btn btn-primary wide">
                              Add
                            </button></div>
                        </div>
                        <p class="text-danger mt-5" style="display: none;">
                          Sum and participations don't match!</p>
                      </div>
                    </form>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <footer class="modal-footer">
            <div class="col-12">
              <div class="row">
                <div class="col"><button type="button" class="btn w-100 btn-primary">Save</button></div>
              </div>
            </div>
          </footer>
        </div>
      </div>
    </div>
    <!---->
    <!---->
  </div>
</div>
`;
